src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Data;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Autofac;
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
```

Comment:
Yeah the problem is that when you write a unit test that fires off multiple `Task` objects that create and commit transactions simultaneously using `ITransctionManager`, the following interesting thing occurs: 
![2015-08-21_14-54-32](https://cloud.githubusercontent.com/assets/938393/9409774/a7286b7c-4814-11e5-9599-c6acecdcdadd.png)

Synchronizing thread access to the transaction manager from the calling code prevents this. The `IDistributedLockService.AcquireLockRecord` creates and commits and disposes a transaction using `ITransactionManager`. I added a unit test called `MultithreadedAcquisitionsShouldNotCauseTransactionErrors` that will fail if you remove the lock.